File: problems/5.string/golang/find_longest_palindrome.go
package main

import (
	"math"
)

func checkPalindrome4(str string) bool {
	R := int(math.Floor(float64(len(str) / 2)))
	L := R
	if len(str)%2 == 0 {
		L = R - 1
	}
	isPalindrome := true

	for R < len(str) && L >= 0 {
		if str[R] != str[L] {
			isPalindrome = false
			break
		}
		R++
		L--
	}

	return isPalindrome
}

func longestPalindrome(str string) string {
	max := ""
	for i := 0; i < len(str); i++ {
		for j := i; j < len(str); j++ {
			if len(str[i:j+1]) > len(max) {
				if checkPalindrome(str[i : j+1]) {
					max = str[i : j+1]
				}
			}
		}
	}

	return max
}

File: ts/5.string/find-longest-palindrome.ts
function checkPalindrome(str: string): boolean {
    if (str.length == 0) return true;
    let R = Math.floor(str.length / 2);
    let L = str.length % 2 === 0 ? R - 1 : R;
    let isPalindrome = true;

    while (R < str.length && L >= 0) {
        if (str[R] !== str[L]) {
            isPalindrome = false;
            break;
        }
        L--;
        R++;
    }

    return isPalindrome;
}

export function longestPalindrome(str: string): string {
    let max = '';
    for (let i = 0; i < str.length; i++) {
        for (let j = i; j < str.length; j++) {
            let temp = str.substring(i, j + 1);
            if (checkPalindrome(temp) && temp.length > max.length) {
                max = temp;
            }
        }
    }

    return max;
}
